src/ripple/app/ledger/impl/TransactionMaster.cpp

```diff
@@ -54,44 +54,57 @@ TransactionMaster::fetch_from_cache(uint256 const& txnID)
     return mCache.fetch(txnID);
 }
 
-std::shared_ptr<Transaction>
+std::optional<std::pair<std::shared_ptr<Transaction>, std::shared_ptr<TxMeta>>>
 TransactionMaster::fetch(uint256 const& txnID, error_code_i& ec)
 {
-    auto txn = fetch_from_cache(txnID);
+    // txn->getLedger() == 0 then txn is not validated and does not contain
+    // metadata
+    if (auto txn = mCache.fetch(txnID); txn && txn->getLedger() == 0)
```

Comment:
There's a possible race condition here. If txn->getLedger() is not set before https://github.com/ripple/rippled/blob/develop/src/ripple/app/ledger/Ledger.cpp#L1010-L1014 then there will be no metadata. I'm not sure if that's the case, but the race condition cannot be introduced.

Reply:
If there is a race condition here, I think it already existed in the code prior to this PR. Previously, the code called txn->getLedger() to get the containing the ledger, and then traversed said ledger to get the metadata. If txn->getLedger() was 0, then no metadata was returned. The intention of this PR is to simply change where we get the metadata from.

Reply:
:+1: